src/engraving/compat/midi/midirender.cpp

```diff
@@ -239,10 +256,17 @@ static void playNote(EventMap* events, const Note* note, PlayNoteParams params,
     }
 
     ev.setVelo(0);
-    if (!note->part()->instrument(note->tick())->useDrumset()
-        && params.offTime != -1) {
+    if (!useDrumSet && params.offTime != -1) {
         events->insert(std::pair<int, NPlayEvent>(std::max(0, params.offTime - params.offset), ev));
     }
+    if (useDrumSet && renderState.isHiHatOpen && (params.pitch == HIHAT_CLOSE_PITCH || params.pitch == HIHAT_PEDAL_PITCH)) {
```

Comment:
maybe we can put all this logic inside RenderState methods (or better some HiHatRenderer)?

Reply:
Not sure it is worth it. This logic only for HH and not going to affect any other events in the future. But RenderState is kind of shared structure for whole render process and could grow in the future to handle other events.